sysml.library/Kernel Library/VectorFunctions.kerml

```diff
@@ -0,0 +1,192 @@
+		inv squareLength { l * l == Inner(v,v) }
+		inv lengthZero { isZero(l) == isZeroVector(v) }
+	}
+	
+	/**
+	 * The angle between two NumericalVectorValues, as a NumericalValue.
+	 */
+	abstract function Angle(v: NumericalVectorValue[1], w: NumericalVectorValue[1]) theta: NumericalValue[1] {
+		inv commutivity { theta == Angle(w, v) }
+		inv lengthInsensitive { theta == Angle(w / Length(w), v / Length(v)) }
+	}
+	
+	/* Specialized functions with concrete definitions for CartesianVectorValues. */
+	
+	/**
+	 * Construct a CartesianVectorValue whose elements are a non-empty list of Real components.
+	 * The dimension of the NumericalVectorValue is equal to the number of components.
+	 */
+	function CartesianVectorOf(components : Real[*]) : CartesianVectorValue {
+		CartesianVectorValue (
+			dimension = size(components),
+			elements = components
+		)
+	}
+	function CartesianThreeVectorOf specializes CartesianVectorOf(components: Real[*]): CartesianThreeVectorValue;
+	
+	/**
+	 * Cartesian zero vectors of 1, 2 and 3 dimensions.
+	 */
+	feature cartesianZeroVector: CartesianVectorValue[3] =
+		(
+			CartesianVectorOf(0.0),
+			CartesianVectorOf((0.0, 0.0)),
+			CartesianThreeVectorOf((0.0, 0.0, 0.0))
+		);
+	feature cartesian3DZeroVector: CartesianThreeVectorValue[1] =
+		cartesianZeroVector[3];
+	
+	/**
+	 * A CartesianVectorValue is a zero vector if all its elements are zero.
+	 */
+	function isCartesianZeroVector specializes isZeroVector (v: CartesianVectorValue): Boolean {
+		v.elements->forAll{in x; x == 0.0}
+	}
+	
+	function 'Cartesian+' specializes '+' (v: CartesianVectorValue[1], w: CartesianVectorValue[0..1]): CartesianVectorValue[1] {
+		inv precondition { w != null implies v.dimension == w.dimension }
+		if w == null? v
+		else CartesianVectorOf(
+			(1..w.dimension)->collect{in i : Positive; v[i] + w[i]}
+		)
+	}
+	
+	function 'Cartesian-' specializes '-' (v: CartesianVectorValue[1], w: CartesianVectorValue[0..1]): CartesianVectorValue[1] {
+		inv precondition { w != null implies v.dimension == w.dimension }
+		CartesianVectorOf(
+			if w == null? CartesianVectorOf(v.elements->collect{in x : Real; -x})
+			else CartesianVectorOf(
+				(1..v.dimension)->collect{in i : Positive; v[i] - w[i]}
+			)
+		)
+	}
+	
+	function CartesianScalarVectorMult specializes ScalarVectorMult (x: Real[1], v: CartesianVectorValue[1]): CartesianVectorValue[1] {
+		CartesianVectorOf(
+			v.elements->collect{in y : Real; x * y}
+		)
+	}
+	function CartesianVectorScalarMult specializes VectorScalarMult (v: CartesianVectorValue[1], x: Real[1]): CartesianVectorValue[1] {
+		CartesianScalarVectorMult(x, v)
+	}
+	
+	function CartesianInner specializes Inner(v: CartesianVectorValue[1], w : CartesianVectorValue[1]): Real[1]{
+		inv precondition { v.dimension == w.dimension }
+		(1..v.dimension)->collect{in i : Positive; v[i] * w[i]}->reduce RealFunctions::'+'
+	}
+	
+	function CartesianLength specializes Length(v: CartesianVectorValue[1]) : NumericalValue[1] {
```

Comment:
CartesianMagnitude ?

Reply:
Updated in commit [3505da2](https://github.com/Systems-Modeling/SysML-v2-Pilot-Implementation/pull/337/commits/3505da240c3983885d275d65d636fde6adcf7ad6).